plugins/usd/module/vtkF3DUSDImporter.cxx

```diff
@@ -0,0 +1,966 @@
+    return this->ConvertMatrix(node.ComputeLocalToWorldTransform(timeCode));
+  }
+
+  void ImportNode(vtkRenderer* renderer, const pxr::UsdPrim& node, const pxr::SdfPath& path,
+    vtkMatrix4x4* currentMatrix)
+  {
+    pxr::UsdTimeCode timeCode = this->CurrentTime * this->Stage->GetTimeCodesPerSecond();
+
+    // simple range-for iteration
+    for (pxr::UsdPrim prim : pxr::UsdPrimSiblingRange(node.GetAllChildren()))
+    {
+      if (prim.IsInstance())
+      {
+        pxr::UsdGeomXform xform = pxr::UsdGeomXform(prim);
+
+        auto mat = this->GetLocalTransform(xform, timeCode);
+
+        this->ImportNode(renderer, prim.GetPrototype(), path.AppendChild(prim.GetName()), mat);
+      }
+      else if (prim.IsA<pxr::UsdGeomPointInstancer>())
+      {
+        pxr::UsdGeomPointInstancer glyphs = pxr::UsdGeomPointInstancer(prim);
+
+        // TODO: Ideally, we should use the 3D glyph mapper, but it's left for future work
+
+        pxr::VtMatrix4dArray xforms;
+
+        if (glyphs.ComputeInstanceTransformsAtTime(&xforms, timeCode, timeCode))
+        {
+          int i = 0;
+          for (const pxr::GfMatrix4d& currInstMatrix : xforms)
+          {
+            auto mat = this->ConvertMatrix(currInstMatrix);
+            vtkMatrix4x4::Multiply4x4(currentMatrix, mat, mat);
+
+            pxr::TfToken tok(std::string("instance_") + std::to_string(i++));
+
+            this->ImportNode(
+              renderer, prim, path.AppendChild(prim.GetName()).AppendChild(tok), mat);
+          }
+        }
+      }
+      else if (prim.IsA<pxr::UsdGeomGprim>())
+      {
+        pxr::UsdGeomGprim geomPrim = pxr::UsdGeomGprim(prim);
+
+        vtkSmartPointer<vtkPolyData> polydata;
+
+        if (prim.IsA<pxr::UsdGeomMesh>())
+        {
+          pxr::UsdGeomMesh meshPrim = pxr::UsdGeomMesh(prim);
+
+          polydata = this->MeshMap[meshPrim.GetPath().GetAsString()];
+          bool meshAlreadyExists = (polydata != nullptr);
+
+          if (!meshAlreadyExists)
+          {
+            polydata = vtkSmartPointer<vtkPolyData>::New();
+          }
+
+          // read points
+          pxr::UsdAttribute pointsAttr = meshPrim.GetPointsAttr();
+
+          if (!meshAlreadyExists || pointsAttr.ValueMightBeTimeVarying())
+          {
+            pxr::VtArray<pxr::GfVec3f> positions;
+            pointsAttr.Get(&positions, timeCode);
+
+            vtkNew<vtkPoints> points;
+            points->Allocate(positions.size());
+            for (const pxr::GfVec3f& p : positions)
+            {
+              points->InsertNextPoint(p[0], p[1], p[2]);
+            }
+
+            polydata->SetPoints(points);
+          }
+
+          // read normals
+          pxr::UsdAttribute normalsAttr = meshPrim.GetNormalsAttr();
+
+          if (meshPrim.GetNormalsInterpolation() == pxr::TfToken("vertex"))
+          {
+            // only vertex interpolation is supported for now
+            // there are many datasets using faceVarying to represent sharp edges
```

Comment:
@mwestphal FYI I'm planning to drop support of face-varying for USD in this PR. I'll log an issue and ideally we need to create a VTKExtension filter to factorize the code between Alembic and USD.

Reply:
please create an issue